### PR TITLE
add TAMP register block for g0, g4, l5, u5 and wl

### PR DESCRIPTION
**Pull request by oll3**
Adds TAMP register block for the devices where I could understand how it was mapped. Some devices seems to have their TAMP/backup/RTC more intertwined and not as a separate register block. Those are not included.

Also not really sure why only u5 and l5 was mapped automatically while I had to add them manually for g0, g4 and wl through extra/families. I guess some field is missing in one of the sources?

Mainly adding this for gaining access to backup registers. Tested on stm32wle5cc.

### CODE REVIEW

**Comment by xoviat** [comment]
Have you looked in the source files to see why chips.rs is not picking it up?


https://embassy.dev/book/dev/developer_stm32.html

**Comment by oll3** [comment]
Seems `InstanceName="TAMP"` is not present in many of the xmls where the corresponding `TAMP` block is present in the svd.
Eg `stm32wle5.svd` has a `TAMP` peripheral but no  `InstanceName="TAMP"`  in `STM32WLE5CCUx.xml`.

How should this be handled?

File: data/registers/tamp_g0.yaml [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+      bit_offset: 19
+      bit_size: 1
+    - name: ITAMP5E
+      description: ITAMP5E
+      bit_offset: 20
+      bit_size: 1
+    - name: ITAMP6E
+      description: ITAMP6E
+      bit_offset: 21
+      bit_size: 1
```
Comment (xoviat) [inline data/registers/tamp_g0.yaml:104]: consolidate to an array

Comment (xoviat) [inline data/registers/tamp_g0.yaml:104]: consolidate to an array

File: data/registers/tamp_g0.yaml [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+      bit_offset: 1
+      bit_size: 1
+    - name: TAMP1MSK
+      description: TAMP1MSK
+      bit_offset: 16
+      bit_size: 1
+    - name: TAMP2MSK
+      description: TAMP2MSK
+      bit_offset: 17
+      bit_size: 1
```
Comment (xoviat) [inline data/registers/tamp_g0.yaml:123]: consolidate to an array

File: data/registers/tamp_g0.yaml [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+  description: control register 2
+  fields:
+    - name: TAMP1NOER
+      description: TAMP1NOER
+      bit_offset: 0
+      bit_size: 1
+    - name: TAMP2NOER
+      description: TAMP2NOER
+      bit_offset: 1
+      bit_size: 1
```
Comment (xoviat) [inline data/registers/tamp_g0.yaml:115]: consolidate to an array

File: data/registers/tamp_g0.yaml [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+      bit_offset: 17
+      bit_size: 1
+    - name: TAMP1TRG
+      description: TAMP1TRG
+      bit_offset: 24
+      bit_size: 1
+    - name: TAMP2TRG
+      description: TAMP2TRG
+      bit_offset: 25
+      bit_size: 1
```
Comment (xoviat) [inline data/registers/tamp_g0.yaml:131]: consolidate to an array

File: data/registers/tamp_g0.yaml [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+      bit_offset: 19
+      bit_size: 1
+    - name: ITAMP5IE
+      description: ITAMP5IE
+      bit_offset: 20
+      bit_size: 1
+    - name: ITAMP6IE
+      description: ITAMP6IE
+      bit_offset: 21
+      bit_size: 1
```
Comment (xoviat) [inline data/registers/tamp_g0.yaml:211]: consolidate to an array

File: data/registers/tamp_g0.yaml [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+      bit_offset: 19
+      bit_size: 1
+    - name: ITAMP5MF
+      description: ITAMP5MF
+      bit_offset: 20
+      bit_size: 1
+    - name: ITAMP6MF
+      description: ITAMP6MF
+      bit_offset: 21
+      bit_size: 1
```
Comment (xoviat) [inline data/registers/tamp_g0.yaml:249]: consolidate to an array

File: data/registers/tamp_g0.yaml [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+      bit_offset: 20
+      bit_size: 1
+    - name: CITAMP6F
+      description: CITAMP6F
+      bit_offset: 21
+      bit_size: 1
+    - name: CITAMP7F
+      description: CITAMP7F
+      bit_offset: 22
+      bit_size: 1
```
Comment (xoviat) [inline data/registers/tamp_g0.yaml:284]: consolidate to an array

File: data/registers/tamp_g0.yaml [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+      bit_offset: 20
+      bit_size: 1
+    - name: ITAMP6F
+      description: ITAMP6F
+      bit_offset: 21
+      bit_size: 1
+    - name: ITAMP7F
+      description: ITAMP7F
+      bit_offset: 22
+      bit_size: 1
```
Comment (xoviat) [inline data/registers/tamp_g0.yaml:326]: consolidate to an array

**Review by xoviat — CHANGES_REQUESTED** [review]
any fields, etc that have a number suffix need to be consolidated to an array. see the hrtim_v1.yaml for examples of this.

**Comment by xoviat** [comment]
> How should this be handled?

If it's not in the XML, you've done the right thing. I will merge after you arrayize the fields, per the review.

**Comment by xoviat** [comment]
> I will merge after you arrayize the fields, per the review.

Also, to be clear I mean all of the files. Not just the one I reviewed.

**Comment by oll3** [comment]
> any fields, etc that have a number suffix need to be consolidated to an array. see the hrtim_v1.yaml for examples of this.

I see, though many of these fields have flags which are not fully sequential. Should they be merged to arrays anyway?

Eg. ITAMPxIE from `tamp_wl.yaml` has flags on bits 18, 20, 21, 23:
```yaml
    - name: ITAMP3IE
      description: Internal tamper 3 interrupt enable
      bit_offset: 18
      bit_size: 1
    - name: ITAMP5IE
      description: Internal tamper 5 interrupt enable
      bit_offset: 20
      bit_size: 1
    - name: ITAMP6IE
      description: Internal tamper 6 interrupt enable
      bit_offset: 21
      bit_size: 1
    - name: ITAMP8IE
      description: Internal tamper 8 interrupt enable
      bit_offset: 23
      bit_size: 1
```

As an array I guess it would look like this, ignoring that offset 19 and 22 are not valid:
```yaml
    - name: ITAMPIE
      description: Internal tamper 3 interrupt enable
      bit_offset: 18
      bit_size: 1
      array:
        len: 6
        stride: 1
```

**Comment by xoviat** [comment]
Use offsets in that case. Again, see hrtim for an example.

**Comment by Dirbaio** [comment]
In this case the bit offsets line up nicely with the numbers, it's just that there's "holes" because some bits are unimplemented. In this case, just use a normal array. it's OK if you can write to the unimplemented bits, these are typically ignored anyway.

only use offsets if the bits really don't line up. For example if there was this:
- ITAMP1IE on bit 8
- ITAMP2IE on bit 9
- ITAMP3IE on bit 16
- ITAMP4IE on bit 17

**Comment by oll3** [comment]
> In this case the bit offsets line up nicely with the numbers, it's just that there's "holes" because some bits are unimplemented. In this case, just use a normal array. it's OK if you can write to the unimplemented bits, these are typically ignored anyway.
> 
> (using offsets would prevent writing to unimplemented bits, but then it no longer holds that TAMPn is at index n-1, which is confusing)
> ...

Makes sense!

I have made repeated fields into arrays and also removed some more 1 bit enums which seemed fairly obvious. And renamed some enums with confusing names (mainly for u5).

Thank you both for taking the time to review and the guidance, really appreciate it! :star2:

**Review by xoviat — APPROVED** [review]

**Comment by xoviat** [comment]
bender run

**Pull request merged**